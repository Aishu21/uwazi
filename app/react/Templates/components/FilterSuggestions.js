import React, {Component, PropTypes} from 'react';
import {connect} from 'react-redux';

export class FilterSuggestions extends Component {

  formatNames(matches) {
    return matches.reduce((names, match) => {
      names.push(match.template);
      return names;
    }, [])
    .join(', ').replace(/(,) (\w* *\w*$)/, ' and $2');
  }

  findSameLabelProperties(label, templates, currentTemplateId) {
    return templates
    .filter((template) => template._id !== currentTemplateId)
    .map((template) => {
      let property = template.properties.find((prop) => {
        return prop.label.toLowerCase() === label.toLowerCase() & prop.filter;
      });

      if (property) {
        return {template: template.name, property};
      }
    })
    .filter((match) => match);
  }


  renderPerfectMatch(matches) {
    return <p className="text-primary">
      <i className="fa fa-thumbs-o-up" aria-hidden="true"></i>&nbsp;
      This property will be used as filter in addition to the same property in <strong>{this.formatNames(matches)}</strong>.
    </p>;
  }

  renderDifferentTypeMatch(matches) {
    let otherType = matches[0].property.type;
    return <p className="text-warning">
      <i className="fa fa-exclamation-triangle" aria-hidden="true"></i>&nbsp;
      This property has the same label as other in <strong>{this.formatNames(matches)}</strong>,
      but not the same type ({otherType}) and won&#39;t be used together for filtering.
    </p>;
  }

  renderDifferentContentMatch(matches) {
    let otherThesauri = this.props.thesauris.find((thesauri) => {
      return thesauri._id === matches[0].property.content;
    }).name;
    return <p className="text-warning">
      <i className="fa fa-exclamation-triangle" aria-hidden="true"></i>&nbsp;
      This property has the same label and type as other in <strong>{this.formatNames(matches)}</strong>,
      but not the same thesauri ({otherThesauri}) and won&#39;t be used together for filtering.
    </p>;
  }

  filterSuggestions() {
    if (!this.props.filter) {
      return;
    }
    let label = this.props.label;
    let type = this.props.type;
    let content = this.props.content;

    let filterMatches = this.findSameLabelProperties(label, this.props.templates, this.props.parentTemplateId);

    let perfectMatches = filterMatches.filter((match) => match.property.type === type && match.property.content === content);

    if (perfectMatches.length) {
      return this.renderPerfectMatch(perfectMatches);
    }

    let differentType = filterMatches.filter((match) => match.property.type !== type);
    if (differentType.length) {
      return this.renderDifferentTypeMatch(differentType);
    }

    let differentContent = filterMatches.filter((match) => match.property.type === type && match.property.content !== content);
    if (differentContent.length) {
      return this.renderDifferentContentMatch(differentContent);
    }
  }

  render() {
    return <div className="filter-suggestions col-sm-12">
            {this.filterSuggestions()}
           </div>;
  }
}

FilterSuggestions.propTypes = {
  label: PropTypes.string,
  type: PropTypes.string,
  filter: PropTypes.any,
  parentTemplateId: PropTypes.string,
  templates: PropTypes.array,
  thesauris: PropTypes.array,
  content: PropTypes.string
};

export function mapStateToProps(state) {
  return {
    templates: state.template.uiState.toJS().templates,
    parentTemplateId: state.template.data.toJS()._id,
    thesauris: state.template.uiState.toJS().thesauris
  };
}

export default connect(mapStateToProps)(FilterSuggestions);
