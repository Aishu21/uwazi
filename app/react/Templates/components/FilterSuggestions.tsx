import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'UI';

import { TemplateSchema } from 'shared/types/templateType';
import { IStore } from 'app/istore';
import { ensure } from 'shared/tsUtils';
import { PropertySchema } from 'shared/types/commonTypes';
import { IImmutable } from 'shared/types/Immutable';
import Icons from './Icons';

const titles = {
  defaultTitle:
    'This property has the same configuration as others with the same label and will be used together.',
  contentConflict:
    'Properties with the same label but different thesauri as content are not allowed.',
  relationConflict:
    'Relationship properties with the same label but different relationship types are not allowed.',
  typeConflict: 'Properties with the same label but different types are not allowed.',
};

interface TemplateProperty {
  template: string;
  relationTypeName: string;
  thesaurusName: string;
  typeConflict: boolean;
  relationConflict: boolean;
  contentConflict: boolean;
  type: string;
  property: PropertySchema;
}

interface MapStateProps {
  templateProperty: TemplateProperty;
}

export const SimilarProperty = (props: MapStateProps) => (
  <tr className="property-atributes is-active">
    <td>
      <Icon icon="file" /> {props.templateProperty.template}
    </td>
    <td
      {...((props.templateProperty.typeConflict || props.templateProperty.relationConflict) && {
        className: 'conflict',
      })}
      {...(props.templateProperty.typeConflict && { title: titles.typeConflict })}
      {...(props.templateProperty.relationConflict && { title: titles.relationConflict })}
    >
      {(props.templateProperty.typeConflict || props.templateProperty.relationConflict) && (
        <Icon icon="exclamation-triangle" />
      )}
      {/*<Icon icon={Icons[props.templateProperty.type.toLowerCase()] || 'fa fa-font'} />*/}
      <Icon icon={Icons.numeric || 'fa fa-font'} />
      {` ${props.templateProperty.type}`}
      {props.templateProperty.relationTypeName && ` (${props.templateProperty.relationTypeName})`}
    </td>
    <td
      className={props.templateProperty.contentConflict ? 'conflict' : ''}
      {...(props.templateProperty.contentConflict && { title: titles.contentConflict })}
    >
      {props.templateProperty.contentConflict && <Icon icon="exclamation-triangle" />}
      {props.templateProperty.thesaurusName && <Icon icon="book" />}
      {props.templateProperty.thesaurusName}
    </td>
  </tr>
);

SimilarProperty.propTypes = {
  templateProperty: PropTypes.object.isRequired,
};

interface MatchedProperty {
  template: string;
  property: PropertySchema;
}

export class FilterSuggestions extends Component<FilterSuggestionsProps> {
  getRelationTypeName(relationTypeId: string) {
    const relationType = relationTypeId
      ? this.props.relationTypes.toJS().find((r: any) => r._id === relationTypeId)
      : undefined;
    return relationType ? relationType.name : undefined;
  }

  getThesauriName(thesauriId: string | undefined) {
    const thesaurus = thesauriId
      ? this.props.thesauris.toJS().find((thesauri: any) => thesauri._id === thesauriId)
      : undefined;
    return thesaurus ? thesaurus.name : undefined;
  }

  findSameLabelProperties(label: string, templates: TemplateSchema[]): MatchedProperty[] {
    return templates
      .filter(template => template._id !== this.props.templateId && template.properties)
      .map(template => {
        const property = ensure<PropertySchema[]>(template.properties).find(
          prop => prop.label.trim().toLowerCase() === label.trim().toLowerCase()
        );
        return (property
          ? { template: template.name, property }
          : { template: undefined }) as MatchedProperty;
      })
      .filter(prop => prop.template !== undefined);
  }

  render() {
    const { label, type, content, relationType } = this.props;

    const similarProperties: TemplateProperty[] = this.findSameLabelProperties(
      this.props.label,
      this.props.templates.toJS()
    ).map((propertyMatch: MatchedProperty) => {
      const { property } = ensure<MatchedProperty>(propertyMatch);
      return Object.assign({}, propertyMatch, {
        typeConflict: property.type !== type,
        relationConflict: relationType && property.relationType !== relationType,
        contentConflict: property.content !== content,
        type: property.type[0].toUpperCase() + property.type.slice(1),
        relationTypeName: this.getRelationTypeName(property.relationType),
        thesaurusName: this.getThesauriName(property.content),
      }) as TemplateProperty;
    });

    const thisProperty = {
      template: `${this.props.templateName} (this template)`,
      property: {
        content,
        label,
        type,
      },
      type: type[0].toUpperCase() + type.slice(1),
      relationTypeName: this.getRelationTypeName(relationType),
      thesaurusName: this.getThesauriName(content),
    };

    const templatesWithSameLabelProperties = [
      thisProperty as TemplateProperty,
      ...similarProperties,
    ];
    const hasContent = templatesWithSameLabelProperties.find(prop => prop.thesaurusName);
    return (
      <>
        <label className="suggestions-label">
          Properties from other templates in the collection using the same label
        </label>
        <table className="table">
          <thead>
            <tr>
              <th>Template</th>
              <th>Type</th>
              {hasContent && <th>Thesauri/Document</th>}
            </tr>
          </thead>
          <tbody>
            {templatesWithSameLabelProperties.map(templateProperty => (
              <SimilarProperty
                key={templateProperty.template + templateProperty.property.name}
                templateProperty={templateProperty}
              />
            ))}
          </tbody>
        </table>
      </>
    );
  }
}

export type FilterSuggestionsProps = {
  index: number;
  label: string;
  type: string;
  filter: any;
  templateName: string;
  templateId: string | { [k: string]: any } | undefined;
  templates: IImmutable<TemplateSchema[]>;
  thesauris: IImmutable<[]>;
  relationTypes: IImmutable<[]>;
  content: string;
  relationType: string;
};

export function mapStateToProps(state: IStore, props: FilterSuggestionsProps) {
  const propertySchemaElement = ensure<PropertySchema>(state.template.data.properties)[props.index];
  // @ts-ignore
  const relationTypes = state.relationTypes;
  return {
    templates: state.templates,
    thesauris: state.thesauris,
    relationTypes: relationTypes,
    templateName: state.template.data.name,
    templateId: state.template.data._id,
    type: propertySchemaElement.type,
    filter: propertySchemaElement.filter,
    label: propertySchemaElement.label,
    content: propertySchemaElement.content,
    relationType: propertySchemaElement.relationType,
  };
}

// @ts-ignore
export default connect(mapStateToProps)(FilterSuggestions);
