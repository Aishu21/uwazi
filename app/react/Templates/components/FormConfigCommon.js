import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Field } from 'react-redux-form';
import { connect } from 'react-redux';
import Tip from 'app/Layout/Tip';

import PrioritySortingLabel from './PrioritySortingLabel';

export class FormConfigCommon extends Component {
  getZeroIndex() {
    const { index, data } = this.props;
    const baseZeroIndex = index + data.commonProperties.length;
    return baseZeroIndex;
  }

  renderTitleField() {
    const { index, formState } = this.props;
    let labelClass = 'form-group';
    const labelKey = `commonProperties.${this.getZeroIndex()}.label`;
    const requiredLabel = formState.$form.errors[`${labelKey}.required`];
    const duplicatedLabel = formState.$form.errors[`${labelKey}.duplicated`];
    if (requiredLabel || duplicatedLabel) {
      labelClass += ' has-error';
    }

    return (
      <div className={labelClass}>
        <label htmlFor={`label${index}`}>Name</label>
        <Field model={`template.data.commonProperties[${this.getZeroIndex()}].label`}>
          <input id={`label${index}`} className="form-control" />
        </Field>
      </div>
    );
  }

  render() {
    const { index, data } = this.props;
    const property = data.commonProperties[this.getZeroIndex()];

    return (
      <div>
        {property.name === 'title' && this.renderTitleField()}
        <Field model={`template.data.commonProperties[${this.getZeroIndex()}].prioritySorting`}>
          <input id={`prioritySorting${index}`} type="checkbox" />
          &nbsp;
          <PrioritySortingLabel htmlFor={`prioritySorting${index}`} />
        </Field>
        <Field model={`template.data.commonProperties[${this.getZeroIndex()}].showInCard`}>
          <input id={`showInCard${index}`} type="checkbox" />
          &nbsp;
          <span>Show in cards</span>
          <Tip>This property will appear in the library cards as part of the basic info.</Tip>
        </Field>
      </div>
    );
  }
}

FormConfigCommon.propTypes = {
  data: PropTypes.shape({ commonProperties: PropTypes.array }).isRequired,
  formState: PropTypes.shape({ $form: PropTypes.object }).isRequired,
  index: PropTypes.number,
};

export function mapStateToProps({ template }) {
  return {
    data: template.data,
    formState: template.formState,
  };
}

export default connect(mapStateToProps)(FormConfigCommon);
