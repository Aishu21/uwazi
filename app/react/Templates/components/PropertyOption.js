import React, {Component, PropTypes} from 'react';
import {DragSource} from 'react-dnd';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import {removeProperty} from 'app/Templates/actions/templateActions';

export class PropertyOption extends Component {
  render() {
    const {connectDragSource} = this.props;
    const {label} = this.props;
    let iconClass = 'fa fa-font';

    if (this.props.type === 'list') {
      iconClass = 'fa fa-list';
    }

    if (this.props.type === 'select') {
      iconClass = 'fa fa-sort';
    }

    if (this.props.type === 'date') {
      iconClass = 'fa fa-calendar';
    }

    if (this.props.type === 'checkbox') {
      iconClass = 'fa fa-check-square-o';
    }

    return (
      connectDragSource(
        <li className="list-group-item" >
          <i className="fa fa-clone"></i>&nbsp;
          <i className={iconClass}></i>&nbsp;{label}
        </li>
      )
    );
  }
}

PropertyOption.propTypes = {
  connectDragSource: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired
};

const optionSource = {
  beginDrag(props) {
    return {label: props.label, type: props.type};
  },

  endDrag(props, monitor) {
    const item = monitor.getItem();
    const dropResult = monitor.getDropResult();
    if (!dropResult) {
      return props.removeProperty(item.index);
    }
  }
};

let dragSource = DragSource('METADATA_OPTION', optionSource, (connector) => ({
  connectDragSource: connector.dragSource()
}))(PropertyOption);

export {dragSource as dragSource};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({removeProperty}, dispatch);
}

export default connect(null, mapDispatchToProps, null, {withRef: true})(dragSource);
