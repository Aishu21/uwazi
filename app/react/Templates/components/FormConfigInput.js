import React, {Component, PropTypes} from 'react';
import FilterSuggestions from 'app/Templates/components/FilterSuggestions';
import {FormField} from 'app/Forms';
import {connect} from 'react-redux';

export class FormConfigInput extends Component {

  validation() {
    return {
      required: (val) => val.trim() !== '',
      duplicated: (val) => {
        return this.props.model.properties.reduce((validity, prop) => {
          let differentLabel = prop.localID === this.props.formKey || prop.label !== val;
          return validity && differentLabel;
        }, true);
      }
    };
  }

  render() {
    const {index, model, formState} = this.props;
    const ptoperty = model.properties[index];
    let labelClass = 'input-group';
    let labelKey = `properties.${index}.label`;
    let duplicatedLabel = formState.fields[labelKey] && formState.fields[labelKey].errors.duplicated;
    if (
      formState.fields[labelKey] &&
      !formState.fields[labelKey].valid &&
      (formState.submitFailed || formState.fields[labelKey].dirty || duplicatedLabel)) {
      labelClass += ' has-error';
    }

    return (
      <div>
        <div className="row">
          <div className="col-sm-4">
            <div className={labelClass}>
              <span className="input-group-addon">
                {(() => {
                  if (duplicatedLabel) {
                    return <span>
                            Duplicated&nbsp;
                            </span>;
                  }
                })()}
                Label
              </span>
              <FormField model={`template.model.properties[${index}].label`} validators={this.validation()}>
                <input className="form-control" />
              </FormField>
            </div>
          </div>
          <div className="col-sm-4">
            <div className="input-group">
              <span className="input-group-addon">
                <FormField model={`template.model.properties[${index}].required`}>
                  <input id={'required' + index} type="checkbox" className="asd"/>
                </FormField>
              </span>
              <label htmlFor={'required' + index} className="form-control">Required field</label>
            </div>
          </div>
        </div>
        <div className="well">
          <div className="row">
            <div className="col-sm-4">
              <FormField model={`template.model.properties[${index}].filter`}>
                <input id={'filter' + this.props.index} type="checkbox"/>
              </FormField>
              &nbsp;
              <label htmlFor={'filter' + this.props.index}>Use as library filter</label>
              <small>This property will be used togheter for filtering with other equal to him.</small>
            </div>
            <div className="col-sm-8">
              <FilterSuggestions {...ptoperty} />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

FormConfigInput.propTypes = {
  model: PropTypes.object,
  index: PropTypes.number,
  formState: PropTypes.object,
  formKey: PropTypes.string
};

export function mapStateToProps({template}) {
  return {
    model: template.model,
    formState: template.formState
  };
}

export default connect(mapStateToProps)(FormConfigInput);
