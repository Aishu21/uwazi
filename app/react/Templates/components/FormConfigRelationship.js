import { Field } from 'react-redux-form';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { Select } from 'app/ReactReduxForms';
import { t } from 'app/I18N';
import FilterSuggestions from 'app/Templates/components/FilterSuggestions';

export class FormConfigRelationship extends Component {
  static contentValidation() {
    return { required: val => val && val.trim() !== '' };
  }

  render() {
    const { index, data, formState } = this.props;
    const thesauris = this.props.thesauris.toJS();
    const property = data.properties[index];
    const relationTypes = this.props.relationTypes.toJS();

    const options = thesauris.filter(thesauri => thesauri._id !== data._id && thesauri.type === 'template');

    let labelClass = 'form-group';
    const labelKey = `properties.${index}.label`;
    const requiredLabel = formState.$form.errors[`${labelKey}.required`];
    const duplicatedLabel = formState.$form.errors[`${labelKey}.duplicated`];
    const relationTypeError = formState.$form.errors[`properties.${index}.relationType.required`] && formState.$form.submitFailed;
    if (requiredLabel || duplicatedLabel) {
      labelClass += ' has-error';
    }

    return (
      <div>
        <div className={labelClass}>
          <label>Label</label>
          <Field model={`template.data.properties[${index}].label`}>
            <input className="form-control"/>
          </Field>
        </div>

        <div className="form-group">
          <label>{t('System', 'Select list')}</label>
          <Select
            model={`template.data.properties[${index}].content`}
            options={options}
            optionsLabel="name"
            placeholder="Any entity or document"
            optionsValue="_id"
          />
        </div>

        <div className={relationTypeError ? 'form-group has-error' : 'form-group'}>
          <label>{t('System', 'Relationship')}<span className="required">*</span></label>
          <Select
            model={`template.data.properties[${index}].relationType`}
            options={relationTypes}
            optionsLabel="name"
            validators={FormConfigRelationship.contentValidation()}
            optionsValue="_id"
          />
        </div>

        <Field model={`template.data.properties[${index}].required`}>
          <input id={`required${this.props.index}`} type="checkbox"/>
          &nbsp;
          <label className="property-label" htmlFor={`required${this.props.index}`}>
            Required property
            <i className="property-help fa fa-question-circle">
              <div className="property-description">You won't be able to publish a document if this property is empty.</div>
            </i>
          </label>
        </Field>

        <Field model={`template.data.properties[${index}].showInCard`}>
          <input id={`showInCard${this.props.index}`} type="checkbox"/>
          &nbsp;
          <label className="property-label" htmlFor={`showInCard${this.props.index}`}>
            Show in cards
            <i className="property-help fa fa-question-circle">
              <div className="property-description">This property will appear in the library cards as part of the basic info.</div>
            </i>
          </label>
        </Field>

        <div>
          <Field model={`template.data.properties[${index}].filter`}>
            <input id={`filter${this.props.index}`} type="checkbox"/>
            &nbsp;
            <label className="property-label" htmlFor={`filter${this.props.index}`}>
              Use as filter
              <i className="property-help fa fa-question-circle">
                <div className="property-description">
                  This property will be used for filtering the library results.
                  When properties match in equal name and field type with other document types, they will be combined for filtering.
                </div>
              </i>
            </label>
          </Field>
          {property.filter && (
            <Field className="filter" model={`template.data.properties[${index}].defaultfilter`}>
              <input
                id={`defaultfilter${this.props.index}`}
                type="checkbox"
                disabled={!property.filter}
              />
              &nbsp;
              <label className="property-label" htmlFor={`defaultfilter${this.props.index}`}>
                Default filter
                <i className="property-help fa fa-question-circle">
                  <div className="property-description">
                    Use this property as a default filter in the library.
                    When there are no document types selected, this property will show as a default filter for your collection.
                  </div>
                </i>
              </label>
            </Field>
          )
          }
          <FilterSuggestions {...property} />
        </div>
      </div>
    );
  }
}

FormConfigRelationship.propTypes = {
  thesauris: PropTypes.object,
  relationTypes: PropTypes.object,
  data: PropTypes.object,
  index: PropTypes.number,
  formState: PropTypes.object
};

export function mapStateToProps(state) {
  return {
    data: state.template.data,
    thesauris: state.thesauris,
    relationTypes: state.relationTypes,
    formState: state.template.formState
  };
}

export default connect(mapStateToProps)(FormConfigRelationship);
