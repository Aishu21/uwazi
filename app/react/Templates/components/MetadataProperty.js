import React, {Component, PropTypes} from 'react';
import {DragSource, DropTarget} from 'react-dnd';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import {editProperty} from 'app/Templates/actions/uiActions';
import {removeProperty, reorderProperty, addProperty} from 'app/Templates/actions/templateActions';
import FormConfigInput from 'app/Templates/components/FormConfigInput';
import FormConfigSelect from 'app/Templates/components/FormConfigSelect';

export class MetadataProperty extends Component {

  renderForm() {
    if (this.props.type === 'select' || this.props.type === 'list') {
      return <FormConfigSelect formKey={this.props.id} index={this.props.index} />;
    }
    return <FormConfigInput formKey={this.props.id} index={this.props.index} />;
  }

  hasError(form) {
    if (Object.keys(form).length === 0) {
      return false;
    }

    let labelError = false;
    if (form.label.touched && !form.label.value) {
      labelError = true;
    }

    let contentError = false;
    let hasContent = this.props.type === 'select' || this.props.type === 'list';
    if (hasContent && form.content.touched && !form.content.value) {
      contentError = true;
    }

    return labelError || contentError;
  }

  render() {
    const {inserting, label, connectDragSource, isDragging, connectDropTarget, editingProperty, index, localID, form} = this.props;
    let propertyClass = 'list-group-item';

    if (this.hasError(form)) {
      propertyClass += ' error';
    }

    if (isDragging || inserting) {
      propertyClass += ' dragging';
    }

    let iconClass = 'fa fa-font';
    if (this.props.type === 'select' || this.props.type === 'list') {
      iconClass = 'fa fa-list';
    }

    if (this.props.type === 'date') {
      iconClass = 'fa fa-calendar';
    }

    if (this.props.type === 'checkbox') {
      iconClass = 'fa fa-check-square-o';
    }

    return connectDragSource(connectDropTarget(
      <li className={propertyClass}>
        <div>
           <i className="fa fa-arrows-v"></i>&nbsp;<i className={iconClass}></i>&nbsp;{label}
          <button className="btn btn-danger btn-xs pull-right property-remove" onClick={() => this.props.removeProperty(index)}>
            <i className="fa fa-trash"></i> Delete
          </button>
          &nbsp;
          <button className="btn btn-default btn-xs pull-right property-edit" onClick={() => this.props.editProperty(localID)}>
            <i className="fa fa-pencil"></i> Edit
          </button>
        </div>
        <div className={'propery-form' + (editingProperty === localID ? ' expand' : '') }>
          {this.renderForm()}
        </div>
      </li>
    ));
  }
}

MetadataProperty.propTypes = {
  connectDragSource: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
  isDragging: PropTypes.bool.isRequired,
  localID: PropTypes.any.isRequired,
  type: PropTypes.string,
  label: PropTypes.string.isRequired,
  inserting: PropTypes.bool,
  removeProperty: PropTypes.func,
  editingProperty: PropTypes.string,
  editProperty: PropTypes.func,
  form: PropTypes.object
};


const target = {
  drop(props) {
    return props;
  },

  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    if (typeof dragIndex === 'undefined') {
      let item = monitor.getItem();
      item.index = 0;
      return props.addProperty({label: item.label, type: item.type, inserting: true}, item.index);
    }

    props.reorderProperty(dragIndex, hoverIndex);
    monitor.getItem().index = hoverIndex;
  }
};

let dropTarget = DropTarget(['METADATA_PROPERTY', 'METADATA_OPTION'], target, (connector) => ({
  connectDropTarget: connector.dropTarget()
}))(MetadataProperty);

const source = {
  beginDrag(props) {
    return {
      index: props.index,
      label: props.label,
      type: props.type
    };
  }
};

let dragSource = DragSource('METADATA_PROPERTY', source, (connector, monitor) => ({
  connectDragSource: connector.dragSource(),
  isDragging: monitor.isDragging()
}))(dropTarget);

function mapDispatchToProps(dispatch) {
  return bindActionCreators({removeProperty, reorderProperty, addProperty, editProperty}, dispatch);
}

const mapStateToProps = (state, props) => {
  return {
    editingProperty: state.template.uiState.toJS().editingProperty,
    form: state.form.template[props.localID] || {}
  };
};

export {dragSource, dropTarget};
export default connect(mapStateToProps, mapDispatchToProps, null, {withRef: true})(dragSource);
