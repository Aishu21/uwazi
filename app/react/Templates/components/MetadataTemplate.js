import React, {PropTypes, Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {DropTarget} from 'react-dnd';
import {Form} from 'react-redux-form';
import {FormField} from 'app/Forms';
import {Link} from 'react-router';

import {inserted, addProperty} from 'app/Templates/actions/templateActions';
import MetadataProperty from 'app/Templates/components/MetadataProperty';
import RemovePropertyConfirm from 'app/Templates/components/RemovePropertyConfirm';

export class MetadataTemplate extends Component {
  render() {
    const {connectDropTarget, isOver} = this.props;

    return connectDropTarget(
      <div>
        <Form model="template.model" className="metadataTemplate panel-default panel">
          <RemovePropertyConfirm />
          <div className="metadataTemplate-heading panel-heading">
            <div className="form-group">
              <FormField model="template.model.name">
                <input placeholder="Template name" className="form-control"/>
              </FormField>
            </div>
            &nbsp;
            <Link to="/metadata" className="btn btn-default"><i className="fa fa-arrow-left"></i> Back</Link>
            &nbsp;
            <button type="submit" className="btn btn-success save-template">
              <i className="fa fa-save"/> Save Template
            </button>
          </div>
          <ul className="metadataTemplate-list list-group">
            {(() => {
              if (this.props.properties.length === 0) {
                return <div className={'no-properties' + (isOver ? ' isOver' : '')}>
                        <i className="fa fa-clone"></i>Drag properties here to start
                      </div>;
              }
            })()}
            {this.props.properties.map((config, index) => {
              return <MetadataProperty {...config} key={config.localID} index={index}/>;
            })}
          </ul>
        </Form>
      </div>
    );
  }
}

MetadataTemplate.propTypes = {
  connectDropTarget: PropTypes.func.isRequired,
  properties: PropTypes.array,
  isOver: PropTypes.bool
};

const target = {
  canDrop() {
    return true;
  },

  drop(props, monitor) {
    let item = monitor.getItem();

    let propertyAlreadyAdded = props.properties[item.index];

    if (propertyAlreadyAdded) {
      props.inserted(item.index);
      return;
    }

    props.addProperty({label: item.label, type: item.type}, props.properties.length);
    return {name: 'container'};
  }
};

let dropTarget = DropTarget('METADATA_OPTION', target, (connector, monitor) => ({
  connectDropTarget: connector.dropTarget(),
  isOver: monitor.isOver()
}))(MetadataTemplate);

export {dropTarget};

const mapStateToProps = ({template}) => {
  return {properties: template.model.properties};
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({inserted, addProperty}, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps, null, {withRef: true})(dropTarget);
