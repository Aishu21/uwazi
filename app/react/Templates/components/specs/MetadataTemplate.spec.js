import React, {Component} from 'react';
import TestUtils from 'react-addons-test-utils';
import TestBackend from 'react-dnd-test-backend';
import {DragDropContext} from 'react-dnd';
import {Provider} from 'react-redux';
import {createStore} from 'redux';
import Immutable from 'immutable';
import {reducer as formReducer} from 'redux-form';

import MetadataTemplate, {MetadataTemplate as DumbComponent, dropTarget} from '~/Templates/components/MetadataTemplate';
import MetadataProperty, {dropTarget as MetadataTarget} from '~/Templates/components/MetadataProperty';
import {dragSource} from '~/Templates/components/PropertyOption';

function sourceTargetTestContext(Target, Source, actions) {
  return DragDropContext(TestBackend)(
    class TestContextContainer extends Component {
      render() {
        const identity = x => x;
        let properties = [{label: 'childTarget', id: 'childId', inserting: true}];
        let targetProps = {properties: properties, connectDropTarget: identity};
        let sourceProps = {label: 'source', index: 2, id: 'source', connectDragSource: identity};
        return <div>
                <Target {...targetProps} {...actions}/>
                <Source {...sourceProps} />
              </div>;
      }
    }
  );
}

function wrapInTestContext(DecoratedComponent) {
  return DragDropContext(TestBackend)(
    class TestContextContainer extends Component {
      render() {
        return <DecoratedComponent {...this.props} />;
      }
    }
  );
}

describe('MetadataTemplate', () => {
  function renderComponent(ComponentToRender, props = {}, properties = []) {
    let result;
    let templateData = Immutable.fromJS({name: '', properties: properties});
    let store = createStore(() => {
      return {
        template: {data: templateData, uiState: Immutable.fromJS({})},
        form: formReducer
      };
    });
    TestUtils.renderIntoDocument(<Provider store={store}><ComponentToRender ref={(ref) => result = ref} {...props}/></Provider>);
    return result;
  }

  let TestComponent;
  beforeEach(() => {
    TestComponent = wrapInTestContext(MetadataTemplate);
  });

  it('should have mapped actions into props', () => {
    let component = renderComponent(TestComponent);
    let template = TestUtils.findRenderedComponentWithType(component, MetadataTemplate).getWrappedInstance();

    expect(template.props.updateProperty).toEqual(jasmine.any(Function));
    expect(template.props.addProperty).toEqual(jasmine.any(Function));
  });

  it('should have mapped store into props', () => {
    let component = renderComponent(TestComponent, null, [{label: 'field', id: 'id2'}]);
    let wrappedComponent = TestUtils.findRenderedComponentWithType(component, MetadataTemplate).getWrappedInstance();

    expect(wrappedComponent.props.properties).toEqual([{label: 'field', id: 'id2'}]);
  });

  describe('render()', () => {
    describe('when fields is empty', () => {
      it('should render a blank state', () => {
        let component = renderComponent(TestComponent, {label: 'test', index: 1, id: 'id'});
        let blankState = TestUtils.findRenderedDOMComponentWithClass(component, 'no-properties');

        expect(blankState.innerHTML).toContain('start');
      });
    });

    it('should add isOver className to the span when isOver', () => {
      let props = {properties: [], isOver: false, connectDropTarget: (x) => x};
      let component = TestUtils.renderIntoDocument(<DumbComponent {...props}/>);
      TestUtils.findRenderedDOMComponentWithClass(component, 'no-properties');

      props = {properties: [], isOver: true, connectDropTarget: (x) => x};
      component = TestUtils.renderIntoDocument(<DumbComponent {...props}/>);
      TestUtils.findRenderedDOMComponentWithClass(component, 'no-properties isOver');
    });

    describe('when has fields', () => {
      it('should render all fields as MetadataProperty', () => {
        let component = renderComponent(TestComponent,
          {properties: []}, [{label: 'property1', id: '1'}, {label: 'property2', id: '2'}]
        );
        let properties = TestUtils.scryRenderedComponentsWithType(component, MetadataProperty);

        expect(properties.length).toBe(2);
      });
    });
  });

  describe('dropTarget', () => {
    let actions = jasmine.createSpyObj(['updateProperty', 'addProperty']);
    let backend;
    let component;

    beforeEach(() => {
      let TestDragAndDropContext = sourceTargetTestContext(dropTarget, dragSource, actions);
      component = renderComponent(TestDragAndDropContext);
      backend = component.getManager().getBackend();
    });

    describe('when droping', () => {
      it('should addField on index 0', () => {
        let target = TestUtils.findRenderedComponentWithType(component, dropTarget);
        let source = TestUtils.findRenderedComponentWithType(component, dragSource);

        backend.simulateBeginDrag([source.getHandlerId()]);
        backend.simulateHover([target.getHandlerId()]);
        backend.simulateDrop();

        expect(actions.addProperty).toHaveBeenCalledWith({label: 'source'}, 0);
      });
    });

    describe('when droping on a child', () => {
      it('should updateProperty removing "inserting" flag on item index', () => {
        let target = TestUtils.findRenderedComponentWithType(component, dropTarget);
        let childTarget = TestUtils.findRenderedComponentWithType(component, MetadataTarget);
        let source = TestUtils.findRenderedComponentWithType(component, dragSource);

        backend.simulateBeginDrag([source.getHandlerId()]);
        backend.simulateHover([target.getHandlerId(), childTarget.getHandlerId()]);
        backend.simulateDrop();

        expect(actions.updateProperty).toHaveBeenCalledWith({label: 'childTarget', id: 'childId', inserting: null}, 0);
      });
    });
  });
});
