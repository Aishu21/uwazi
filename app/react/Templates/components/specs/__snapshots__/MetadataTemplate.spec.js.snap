// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataTemplate render() should render template color field 1`] = `
<DefaultConnectedControl
  component={[Function]}
  defaultValue="#112233"
  mapProps={
    Object {
      "defaultValue": [Function],
    }
  }
  model=".color"
/>
`;
